[wei-song1] iP

wei-song1 opened the pull request:
# Codsworth
> The best companion for mankind 😉
```
________________________________________________________
|   _____          _                        _   _      |
|  / ____|        | |                      | | | |     |
| | |     ___   __| |_____      _____  _ __| |_| |__   |
| | |    / _ \ / _` / __\ \ /\ / / _ \| '__| __| '_ \  |
| | |___| (_) | (_| \__ \\ V  V / (_) | |  | |_| | | | |
|  \_____\___/ \__,_|___/ \_/\_/ \___/|_|   \__|_| |_| |
|                                                      |
| Hello! What can I do for you today, master?          |
________________________________________________________
```
* Text based
* Easy to learn
* ~~Better than dog~~

**Features**
- [ ] Remembers your tasks
- [ ] Marks them as completed or not
- [ ] Even keeps track of dates

_All you need to do is,_
1. Download the JAR file [here](https://github.com/wei-song1/ip/releases/tag/A-Jar)
2. Run it
3. Have fun knowing that the future is here!

> [!CAUTION]
> Codsworth is not liable for missed tasks due to user's negligence

This is how you write `Hello World` in Java, because Codsworth loves to say hello to you too!
```
public class Main {
    public static void main(String[] args) {
        System.out.println("Hello World");
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

wei-song1 committed:
Renamed Duke.java into Codsworth.java

wei-song1 committed:
Modified Codsworth.java to Echo

wei-song1 committed:
Added the ability to store whatever text entered by the user and display them back to the user when requested

wei-song1 committed:
Implementation of Level 3 without Task class

wei-song1 committed:
Added Mark to Done and Unmark to Undone, split Codsworth.java into Task.java

wei-song1 committed:
Added support for tracking three types of tasks: ToDos, Deadlines, Events

wei-song1 committed:
Testing using the I/O redirection technique

wei-song1 committed:
Added support for deleting tasks from the list.

wei-song1 committed:
Used Java Enumerations

wei-song1 committed:
Started work for custom exceptions

wei-song1 committed:
Bugs fixed for mark/unmark/delete/todo/deadline/event * Typing just “mark”, “unmark”, “delete”, “todo”, “deadline”, “event” does not throw an exception and close instantly
* Mark/unmark/delete now requires an integer that’s within bounds
* Todo/deadline/event now requires an input/the correct input to be accepted
* If it’s not an actual command, it gives an error with all possible commands

wei-song1 committed:
Fix Codsworth.java switch case and other inconsistencies to have correct indentation

wei-song1 committed:
Add first implementation of file saving that only saves new tasks and adds them back to list upon startup

wei-song1 committed:
Add the ability to keep track if task is done or not, and ability to reset the file

wei-song1 committed:
Merge branch 'branch-Level-7'

wei-song1 committed:
Add function that converts text input into the correct data format for localDate/localDateTime

wei-song1 committed:
Enforce deadline to use the correct date formats

wei-song1 committed:
Change the way deadlines/events dates are stored, now requires it to be formatted correctly, and modify the invalid date exception to be more explicit

wei-song1 committed:
Merge branch 'branch-Level-8'

wei-song1 committed:
Extract all the Ui elements from Codsworth.java into Ui.java

wei-song1 committed:
Extract all the TaskList and Storage into their own TaskList.java and Storage.java files

wei-song1 committed:
Extract all the Parser commands into it's own Parser.java and modify the main Codsworth.java to reflect the new OOP

wei-song1 committed:
Merge branch 'branch-A-MoreOOP'

wei-song1 committed:
Move all the files into package Codsworth, and all the tasks into package Task

wei-song1 committed:
Merge branch 'branch-A-Packages'

wei-song1 committed:
Merge branch 'add-gradle-support'

wei-song1 committed:
Add Gradle and set up checkstyle for future use

wei-song1 committed:
Fix the code to match checkstyle

wei-song1 committed:
Add TaskList JUnit test

wei-song1 committed:
Add Parser JUnit test

wei-song1 committed:
Set-up gradle.build to build JAR files

wei-song1 committed:
Update README.md

wei-song1 committed:
Add Javadocs onto Deadline, Event, Task, ToDo and TaskList

wei-song1 committed:
Fix package names of Codsworth, CodsworthException and add javadocs for all remaining classes

wei-song1 committed:
Add javadocs for test cases

wei-song1 committed:
Clean up TaskList, fix other minor issue, and give reset a proper message

wei-song1 committed:
Add the ability to search for matching strings

wei-song1 committed:
Merge branch 'branch-A-JavaDoc'

wei-song1 committed:
Merge branch 'branch-A-CodingStandard'

wei-song1 committed:
Merge branch 'branch-Level-9'

# Conflicts:
#	src/main/java/codsworth/Ui.java

wei-song1 committed:
Commit for check style

wei-song1 committed:
Remove logo and line breaks in preparation for text based bubble non monospace chat

cerulyean commented on src/main/java/codsworth/Task/Task.java:
```diff
@@ -0,0 +1,44 @@
+/**
+ * Represents task. A <code>Task</code> object corresponds to a task represented by a name
+ */
+public class Task {
```
Since Task is a base class and shouldn't be instantiated directly, have you considered making it abstract.

ZaydM18 replied:
The Task class is never used directly. Making it abstract would likely be better.

cerulyean commented on src/test/java/codsworth/Task/TaskListTest.java:
```diff
@@ -0,0 +1,165 @@
+        assertThrows(CodsworthInvalidDateException.class, () -> TaskList.formatCorrectDate(null));
+        assertThrows(CodsworthInvalidDateException.class, () -> TaskList.formatCorrectDate("a"));
+        assertThrows(CodsworthInvalidDateException.class, () -> TaskList.formatCorrectDate("01-01-2001 abcd"));
+        assertThrows(CodsworthInvalidDateException.class, () -> TaskList.formatCorrectDate("ab-cd-efgh 0123"));
```
I like how this accounts for a very large variety of possible date time formats, it looks very flexible

cerulyean commented on src/main/java/codsworth/Parser.java:
```diff
@@ -0,0 +1,99 @@
+import codsworth.codsworthexceptions.CodsworthInvalidDateException;
+import codsworth.codsworthexceptions.CodsworthMissingInputException;
+import codsworth.codsworthexceptions.CodsworthOutOfBoundsException;
+import codsworth.codsworthexceptions.CodsworthWrongFormatException;
```
I like the number of different exceptions, looks like it would be very helpful in maintaining the code and for telling the user what went wrong

cerulyean commented on src/main/java/codsworth/Task/Task.java:
```diff
@@ -0,0 +1,44 @@
+
+    public void setDoneOrUndone(String type) {
+        this.isDone = type.equals("mark");
+    }
```
I noticed that the setDoneOrUndone method in the Task class takes a String parameter ("mark" or "unmark") to determine the task’s status. According to coding standards, methods should avoid using strings for control logic, as this can lead to errors and reduce code clarity. Maybe do this under Parser and use a boolean when calling Task methods instead?

cerulyean commented on src/main/java/codsworth/Task/TaskList.java:
```diff
@@ -0,0 +1,265 @@
+            System.out.println(Ui.getNewTask(temp, taskList.size()));
+
+        // Exception
+        } catch (CodsworthMissingInputException e) {
+            throw new CodsworthMissingInputException();
+        } catch (ArrayIndexOutOfBoundsException e) {
+            throw new CodsworthInvalidDateException();
+        }
```
Catching an exception and throwing a different one seems like it would hide information that is useful. I feel adding more context or just propagating the original exception would be more helpful.

cerulyean reviewed (commented):
Your code looks really well organized and neat, with consistently good and descriptive names for methods and variables. The javadoc comments are also descriptive, clear and through. One relatively minor issue(?) I found is you only calling Storage.saveTaskList() when closing the program, but Task-7 states that saving on TaskList modification. Overall, it looks great.

ZaydM18 commented on src/main/java/codsworth/Parser.java:
```diff
@@ -0,0 +1,99 @@
+            Storage.saveTaskList();
+            break;
+
+        case "reset":
```
Good choice to add a reset button as the project evolved to include a saved task list.

ZaydM18 commented on src/main/java/codsworth/CodsworthExceptions/CodsworthMissingInputException.java:
```diff
@@ -0,0 +1,11 @@
+/**
+ * Thrown when there are missing inputs given to Codsworth
+ */
+public class CodsworthMissingInputException extends RuntimeException {
```
Good use of a more complex exception instead of inheriting the base exception class.

ZaydM18 commented on src/main/java/codsworth/Task/TaskList.java:
```diff
@@ -0,0 +1,265 @@
+     * @param input Task name with its optional dates.
+     * @param operation Operation currently being added.
+     */
+    public static void addTaskAndPrint(String input, String operation) {
```
Seems like this could be compartmentalised into a function for the Parser class instead. Since it looks more like it has to do with instantiation.

ZaydM18 commented on src/main/java/codsworth/Task/TaskList.java:
```diff
@@ -0,0 +1,265 @@
+/**
+ * Represents ArrayList of Task in a TaskList class and encapsulates all it's methods into the class.
+ */
+public class TaskList {
```
The extra TaskList class is useful to have, but some of its functions may be performed better by either the Task or Parser classes.

ZaydM18 reviewed (commented):
Overall, well-designed code that adheres to OOP principles and demonstrates a good understanding of the core topics. Most issues are minor and can be fixed easily. Some structural modifications may be useful but not necessary. Nothing to worry about, it's no Fallout 76.

wei-song1 committed:
Add gui to Codsworth, old code still left untouched

wei-song1 committed:
Add theming to the gui

wei-song1 committed:
Improve theming and make Task into abstract class

wei-song1 committed:
Merge branch 'branch-Level-10'

wei-song1 committed:
Make file save after every command instead of only at bye, clean up old legacy code and remove static references, rename Ui package to ui

wei-song1 committed:
Add assert feature to document important assumptions

wei-song1 committed:
Improve code quality

Instead of throwing and having multiple catch, all catch
statements have been modified to catch only Codsworth exceptions.

Consolidated TaskList add task method so there's no more dupes.

Added more test cases after code quality check unearthed
some rare bugs.

wei-song1 committed:
Merge pull request #1 from wei-song1/branch-A-Assertions

Add assert feature to document important assumptions

wei-song1 committed:
Merge pull request #2 from wei-song1/branch-A-CodeQuality

Improve code quality

wei-song1 committed:
Add 2 BCD extensions

* Add C-Help, help is now a command
* Add C-Duplicate, duplicate commands now gets exception
* Add CodsworthDuplicateException
* Remove useless import from all task/subtasks

wei-song1 committed:
Make GUI more consistent

wei-song1 committed:
Add more test and make program exit when file cannot be loaded

wei-song1 committed:
Fix timing from being save wrongly

wei-song1 committed:
Add user guide

wei-song1 committed:
Attempt to fix Ui picture

wei-song1 committed:
Fix weak slap
